pkg/quota/admission/clusterresourceoverride/admission.go

```diff
@@ -0,0 +1,161 @@
+	return config, err
+}
+
+func (a *clusterResourceOverridePlugin) Validate() error {
+	if err := validate.Validate(a.Config); err != nil {
+		return err
+	}
+	if a.ProjectCache == nil {
+		return fmt.Errorf("ClusterResourceOverride did not get a project cache")
+	}
+	return nil
+}
+
+// TODO this will need to update when we have pod requests/limits
+func (a *clusterResourceOverridePlugin) Admit(attr admission.Attributes) error {
+	glog.V(8).Infof("ClusterResourceOverride admission controller is invoked")
+	if a.Config == nil || attr.GetResource() != kapi.Resource("pods") || attr.GetSubresource() != "" {
+		return nil // not applicable
+	}
+	pod, ok := attr.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+	glog.V(5).Infof("ClusterResourceOverride is looking at creating pod %s in project %s", pod.Name, attr.GetNamespace())
+
+	// allow annotations on project to override
+	if ns, err := a.ProjectCache.GetNamespace(attr.GetNamespace()); err != nil {
+		glog.Warningf("ClusterResourceOverride got an error retrieving namespace: %v", err)
+		return admission.NewForbidden(attr, err) // this should not happen though
+	} else {
+		projectEnabledPlugin, exists := ns.Annotations[clusterResourceOverrideAnnotation]
+		if exists && projectEnabledPlugin != "true" {
+			glog.V(5).Infof("ClusterResourceOverride is disabled for project %s", attr.GetNamespace())
+			return nil // disabled for this project, do nothing
+		}
+	}
+
+	// Reuse LimitRanger logic to apply limit/req defaults from the project. Ignore validation
+	// errors, assume that LimitRanger will run after this plugin to validate.
+	glog.V(5).Infof("ClusterResourceOverride: initial pod limits are: %#v", pod.Spec.Containers[0].Resources)
+	if err := a.LimitRanger.Admit(attr); err != nil {
+		glog.V(5).Infof("ClusterResourceOverride: error from LimitRanger: %#v", err)
+	}
+	pod, ok = attr.GetObject().(*kapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected object: %#v", attr.GetObject()))
+	}
+	glog.V(5).Infof("ClusterResourceOverride: pod limits after LimitRanger are: %#v", pod.Spec.Containers[0].Resources)
+	for _, container := range pod.Spec.Containers {
+		resources := container.Resources
+		memLimit := resources.Limits.Memory()
+		// Although resource Quantity objects are used for both memory and CPU they're at a very
+		// different scale. The inf.Dec number type is for decimal numbers down to 0.001,
+		// stored as ints with a scale. Memory is measured in bytes, while CPU is in cores,
+		// with 3 places of accuracy for millicores. In order to maintain 3 places of accuracy
+		// for CPU there is some hocus-pocus with the "scale" component when multiplying
+		// (inf.Dec does not have a multiplication method).
+		if a.Config.LimitCPUToMemoryPercent != 0.0 {
+			resources.Limits[kapi.ResourceCPU] = resource.Quantity{
```

Comment:
On Thu, Feb 11, 2016 at 1:22 PM, Derek Carr <notifications@github.com>
wrote:

> This has problems if the other things that do defaulting did not give
> value (i.e. InitialResources or LimitRange).
>

For the case that a project has no LimitRange defaults,  I changed the
logic not to override based on a missing value.

Either this plug-in can do some defaulting itself, or it should reject if
> it sees a pod that gave no memory limit. I think I prefer rejection.
>

At the moment, at least, if limits are specified in a LimitRange, then so
are defaults. So it seems to me the only scenario where this would happen
is when there aren't any limits set; that might be admin error but it seems
strange to me to reject for missing pod limits when there aren't any limits
in a LimitRange to be violated.